synedrion/src/curve/arithmetic.rs

```diff
@@ -1,132 +1,169 @@
+        }
+
+        let generator_bytes = <Self as CurveArithmetic>::ProjectivePoint::generator().to_bytes();
+        digest.chain::<&[u8]>(&generator_bytes.as_ref())
+    }
+}
+
+#[cfg(test)]
+impl HashableType for TinyCurve32 {
+    fn chain_type<C: Chain>(digest: C) -> C {
+        let mut digest = digest;
+        // TODO: see the k256 implementation above.
+        let words = Self::ORDER.to_words();
+        for word in words {
+            digest = digest.chain(&word.to_le_bytes());
+        }
+
+        let generator_bytes = <Self as CurveArithmetic>::ProjectivePoint::generator().to_bytes();
+        digest.chain::<&[u8]>(&generator_bytes.as_ref())
+    }
+}
+
+type BackendScalar<P> = <<P as SchemeParams>::Curve as CurveArithmetic>::Scalar;
+
 #[derive(Clone, Copy, Debug, PartialEq, Eq, Default, PartialOrd, Ord, Zeroize)]
-pub(crate) struct Scalar(BackendScalar);
+pub(crate) struct Scalar<P: SchemeParams>(BackendScalar<P>);
 
-impl Scalar {
-    pub const ZERO: Self = Self(BackendScalar::ZERO);
-    pub const ONE: Self = Self(BackendScalar::ONE);
+impl<P: SchemeParams> Scalar<P> {
+    pub const ZERO: Self = Self(BackendScalar::<P>::ZERO);
+    pub const ONE: Self = Self(BackendScalar::<P>::ONE);
 
     pub fn random(rng: &mut impl CryptoRngCore) -> Self {
-        Self(BackendScalar::random(rng))
+        Self(ScalarPrimitive::<P::Curve>::random(rng).into())
     }
 
     pub fn random_nonzero(rng: &mut impl CryptoRngCore) -> Self {
-        Self(*NonZeroScalar::<Secp256k1>::random(rng).as_ref())
+        Self(*NonZeroScalar::<P::Curve>::random(rng).as_ref())
     }
 
-    pub fn mul_by_generator(&self) -> Point {
-        Point::GENERATOR * self
+    pub fn mul_by_generator(&self) -> Point<P> {
+        Point::generator() * self
     }
 
     /// Invert the [`Scalar`]. Returns [`None`] if the scalar is zero.
     pub fn invert(&self) -> CtOption<Self> {
         self.0.invert().map(Self)
     }
 
+    /// Read twice the number of bytes in a curve [`Scalar`] from the [`XofReader`], then reduce
+    /// modulo the curve order to ensure a valid, unbiased scalar.
     pub fn from_xof_reader(reader: &mut impl XofReader) -> Self {
-        let mut bytes = k256::WideBytes::default();
-        reader.read(&mut bytes);
-        Self(<BackendScalar as Reduce<U512>>::reduce_bytes(&bytes))
+        let bytes_lo = reader.read_boxed(Self::repr_len());
+        let bytes_lo = GenericArray::<_, FieldBytesSize<P::Curve>>::from_slice(&bytes_lo);
+        let bytes_hi = reader.read_boxed(Self::repr_len());
+        let bytes_hi = GenericArray::<_, FieldBytesSize<P::Curve>>::from_slice(&bytes_hi);
+        let uint_lo = <P::Curve as Curve>::Uint::decode_field_bytes(bytes_lo);
+        let uint_hi = <P::Curve as Curve>::Uint::decode_field_bytes(bytes_hi);
+        let wide_uint = uint_lo.concat(&uint_hi);
+        // TODO: When the elliptic curve stack upgrades to crypto-bigint v0.6 we can use RemMixed and
```

Comment:
You can use `crypto-bigint 0.6` (the one we use for Paillier, for example) all the way and in the last line convert it to the old `Uint`, like in `params::convert_uint()`